atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/SweepTaskRunner.java

```diff
@@ -188,32 +190,67 @@ private SweepResults doRun(TableReference tableRef,
         }
     }
 
+    /**
+     * Returns batches with at least batchConfig.deleteBatchSize blocks per batch.
+     */
     private Iterator<BatchOfCellsToSweep> getBatchesToSweep(Iterator<List<CandidateCellForSweeping>> candidates,
                                                             SweepBatchConfig batchConfig,
                                                             SweepableCellFilter sweepableCellFilter,
                                                             ExaminedCellLimit limit) {
+        Iterator<List<CandidateCellForSweeping>> validCandidates =
+                Iterators.filter(candidates, list -> list != null && !list.isEmpty());
+
         Iterator<BatchOfCellsToSweep> cellsToSweep = Iterators.transform(
-                Iterators.filter(candidates, list -> !list.isEmpty()),
+                validCandidates,
                 sweepableCellFilter::getCellsToSweep);
         return new CellsToSweepPartitioningIterator(cellsToSweep, batchConfig.deleteBatchSize(), limit);
     }
 
     /**
-     * Returns the number of deleted (cell, timestamp) pairs.
+     * Partitions batches of blocks down to @param deleteBatchSize blocks per sweep deletion call.
+     * Returns the number of blocks that were deleted.
      */
-    private int sweepBatch(TableReference tableRef, List<CellToSweep> batch, RunType runType) {
-        Multimap<Cell, Long> startTimestampsToSweepPerCell = ArrayListMultimap.create();
-        List<Cell> sentinels = Lists.newArrayList();
+    private int sweepBatch(TableReference tableRef, List<CellToSweep> batch, RunType runType,
+            int deleteBatchSize) {
+        int numberOfSweptCells = 0;
+
+        Multimap<Cell, Long> currentBatch = ArrayListMultimap.create();
+        List<Cell> currentBatchSentinels = Lists.newArrayList();
+
         for (CellToSweep cell : batch) {
-            startTimestampsToSweepPerCell.putAll(cell.cell(), TDecorators.wrap(cell.sortedTimestamps()));
             if (cell.needsSentinel()) {
-                sentinels.add(cell.cell());
+                currentBatchSentinels.add(cell.cell());
+            }
+
+            if (currentBatch.size() + cell.sortedTimestamps().size() < deleteBatchSize) {
+                currentBatch.putAll(cell.cell(), TDecorators.wrap(cell.sortedTimestamps()));
+                continue;
             }
+
+            TLongList remainingCellTimestamps = cell.sortedTimestamps();
+            while (currentBatch.size() + remainingCellTimestamps.size() >= deleteBatchSize) {
+                int numberOfTimestampsForThisBatch = deleteBatchSize - currentBatch.size();
+                TLongList timestampsForThisBatch = remainingCellTimestamps.subList(0, numberOfTimestampsForThisBatch);
```

Comment:
I've updated this code to use ImmutableList. Let me know if I can improve perf in any other way.